scripts/builtin/hmm.dml

```diff
@@ -0,0 +1,258 @@
+    seed = seed+1
+    B = col_normalize(matrix(1/nr_outputs,rows=nr_states, cols=nr_outputs) + rand(rows=nr_states, cols=nr_outputs, seed=seed))
+    seed = seed+1
+    ip = matrix(1/nr_states, rows=nr_states, cols=1) + rand(rows=nr_states, cols=1, seed=seed)
+    seed = seed+1
+
+    likelihood = 0
+    T = ncol(X)
+    
+    for (i in 1:IT) {
+        alpha = forward(X, A, B, ip)
+        beta = backward(X, A, B)
+        
+        gamma = calculate_gamma(alpha, beta)
+        eta = calculate_eta(X, alpha, beta, A, B)
+
+        ip = gamma[, 1]
+        A = calculate_A(eta, gamma, nr_states)
+        B = calculate_B(gamma, X, nr_states, nr_outputs)
+        
+        #likelhood using forward method
+        likelihood = sum(gamma[, T] * alpha[, T])
+    }
+}
+
+calculate_A = function(Matrix[Double] eta, Matrix[Double] gamma, Integer nr_states) return (Matrix[Double] A) {
+    A = rand(rows=nr_states, cols=nr_states)
+    T = ncol(eta)
+
+    parfor (i in 1:nr_states) {
+        parfor (j in 1:nr_states) {
+            eta_id = (i-1) * nr_states + j
+            num_ij = sum(eta[eta_id, 1:T-1])
+            den_ij = sum(gamma[i,1:T-1])
+            A[i,j] = num_ij/den_ij
+        }
+    }
+}
+
+calculate_B = function(Matrix[Double] gamma, Matrix[Double] X, Integer nr_states, Integer nr_outputs) return (Matrix[Double] B) {
+    B = rand(rows=nr_states, cols=nr_outputs)
+
+    parfor (i in 1:nr_states){
+        parfor (j in 1:nr_outputs) {
+            B[i, j] = sum((X == j) * gamma[i, ]) / sum(gamma[i, ])
+        }
+    }
+}
+
+/*
+  To work with 3d matrices, the values in first two dimensions of eta matrix
+  must be rolled the first dimension. The following indexing function maps the
+  index of rolled matrix(eta) to "normal" 2d matrices like alpha, beta, etc
+*/
+index = function(Integer rolled_i, Integer nr_states) return(Double i, Double j) {
+    i = ceil(rolled_i / nr_states) #index starts at 1
+    j = rolled_i - (nr_states * (i-1)) #i-1) * 
+}
+
+calculate_eta =  function (Matrix[Double] X, Matrix[Double] alpha, Matrix[Double] beta, Matrix[Double] A, Matrix[Double] B) return (Matrix[Double] eta)
+{
+    nr_states = nrow(alpha)
+    T = ncol(alpha)    
+    
+    /*
+    eta a (nr_states * nr_states) * T matrix with cell (i, t) being probability of 
+    the state being at i at timestep j given the observed output
+    */
+    
+    tot_transitions = nr_states * nr_states
+    eta = rand(rows=tot_transitions, cols=T-1)
+    
+    /*
+    The transitions will be indiced as such:transition 1-N will represent
+    transition from state 1 to 1, 2 upto state N. transition N+1-2N will represent
+    transitions from 2 to 1, 2 upto state N  
+    */
+    
+    parfor (trans_id in 1:tot_transitions) {
+        [i, j] = index(trans_id, nr_states)
+        for (t in 1:(T-1)) {
+            #indices for alpha and beta
+            ot1 = output_t(X, t+1)
```

Comment:
remove the method output_t and call the extraction here. It should automatically know it is a scalar.
Do you cast to int because you want to round? if so then call round()